tls/s2n_config.c

```diff
@@ -51,109 +46,158 @@ int get_nanoseconds_since_epoch(void *data, uint64_t * nanoseconds)
+
+    return 0;
 }
 
-/* Accept all RSA Certificates is unsafe and is only used in the s2n Client for testing purposes */
-s2n_cert_validation_code accept_all_rsa_certs(struct s2n_connection *conn,
-                                              uint8_t *cert_chain_in,
-                                              uint32_t cert_chain_len,
-                                              struct s2n_cert_public_key *public_key_out,
-                                              void *context)
-{
-    struct s2n_blob cert_chain_blob = { .data = cert_chain_in, .size = cert_chain_len};
-    struct s2n_stuffer cert_chain_in_stuffer;
-    if (s2n_stuffer_init(&cert_chain_in_stuffer, &cert_chain_blob) < 0) {
-        return S2N_CERT_ERR_INVALID;
-    }
-    if (s2n_stuffer_write(&cert_chain_in_stuffer, &cert_chain_blob) < 0) {
-        return S2N_CERT_ERR_INVALID;
+static uint8_t default_verify_host(const char *host_name, size_t len, void *data) {
+    return 1;
+}
+
+#endif
+
+static uint8_t default_config_init = 0;
+static uint8_t unsafe_client_testing_config_init = 0;
+static uint8_t default_client_config_init = 0;
+static uint8_t default_fips_config_init = 0;
+
+
+static struct s2n_config s2n_default_config;
+
+/* This config should only used by the s2n_client for unit/integration testing purposes. */
+static struct s2n_config s2n_unsafe_client_testing_config;
+
+static struct s2n_config default_client_config;
+
+static struct s2n_config s2n_default_fips_config;
+
+static int s2n_config_init(struct s2n_config *config) {
+    config->cert_and_key_pairs = NULL;
+    config->dhparams = NULL;
+    memset(&config->application_protocols, 0, sizeof(config->application_protocols));
+    config->status_request_type = S2N_STATUS_REQUEST_NONE;
+    config->wall_clock = wall_clock;
+    config->monotonic_clock = monotonic_clock;
+    config->verify_host = default_verify_host;
+    config->data_for_verify_host = NULL;
+    config->client_hello_cb = NULL;
+    config->client_hello_cb_ctx = NULL;
+    config->cache_store = NULL;
+    config->cache_store_data = NULL;
+    config->cache_retrieve = NULL;
+    config->cache_retrieve_data = NULL;
+    config->cache_delete = NULL;
+    config->cache_delete_data = NULL;
+    config->ct_type = S2N_CT_SUPPORT_NONE;
+    config->mfl_code = S2N_TLS_MAX_FRAG_LEN_EXT_NONE;
+    config->accept_mfl = 0;
+
+    /* By default, only the client will authenticate the Server's Certificate. The Server does not request or
+     * authenticate any client certificates. */
+    config->client_cert_auth_type = S2N_CERT_AUTH_NONE;
+    config->check_ocsp = 1;
+
+    if (s2n_is_in_fips_mode()) {
+        s2n_config_set_cipher_preferences(config, "default_fips");
+    } else {
+        s2n_config_set_cipher_preferences(config, "default");
     }
 
-    uint32_t certificate_count = 0;
-    while (s2n_stuffer_data_available(&cert_chain_in_stuffer)) {
-        uint32_t certificate_size;
+    s2n_x509_trust_store_init(&config->trust_store);
```

Comment:
We need to ensure the init() function is called on the config structure instead of just returning a struct (since config isn't really POD, it doesn't make sense to treat it like one). Instead of having different ways of initializing the data everywhere, forcing everything to use the init function seems the simplest. 

I moved it to functions so that everything get's init called.